File: Languages/python/algorithms-dsa/leetcode/stack/daily_temperatures.py
"""
LeetCode #739 - Daily Temperatures
Topic: Stack
Difficulty: Medium

Find days until warmer temperature using a monotonic decreasing stack.

Time Complexity: O(n)
Space Complexity: O(n)
"""


def daily_temperatures(temperatures: list[int]) -> list[int]:
    n = len(temperatures)
    result = [0] * n
    stack: list[int] = []  # indices
    for i in range(n):
        while stack and temperatures[i] > temperatures[stack[-1]]:
            idx = stack.pop()
            result[idx] = i - idx
        stack.append(i)
    return result


if __name__ == "__main__":
    assert daily_temperatures([73, 74, 75, 71, 69, 72, 76, 73]) == [1, 1, 4, 2, 1, 1, 0, 0]
    assert daily_temperatures([30, 40, 50, 60]) == [1, 1, 1, 0]
    assert daily_temperatures([30, 60, 90]) == [1, 1, 0]
    print("All tests passed!")

File: Languages/go/algorithms-dsa/leetcode/stack/daily_temperatures.go
/*
LeetCode #739: Daily Temperatures
Topic: Stack
Difficulty: Medium

Given an array of daily temperatures, return an array where answer[i]
is the number of days until a warmer temperature. If no future warmer
day exists, answer[i] == 0.

Approach: Monotonic decreasing stack of indices. When current temp is
warmer than stack top, pop and record the difference in days.

Time: O(n), Space: O(n)
*/
package main

import "fmt"

func dailyTemperatures(temperatures []int) []int {
	n := len(temperatures)
	result := make([]int, n)
	stack := []int{}

	for i := 0; i < n; i++ {
		for len(stack) > 0 && temperatures[i] > temperatures[stack[len(stack)-1]] {
			idx := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			result[idx] = i - idx
		}
		stack = append(stack, i)
	}
	return result
}

func main() {
	r := dailyTemperatures([]int{73, 74, 75, 71, 69, 72, 76, 73})
	exp := []int{1, 1, 4, 2, 1, 1, 0, 0}
	pass := true
	for i := range r {
		if r[i] != exp[i] {
			pass = false
		}
	}
	if pass {
		fmt.Println("PASS Test 1")
	} else {
		fmt.Println("FAIL Test 1")
	}

	r2 := dailyTemperatures([]int{30, 40, 50, 60})
	exp2 := []int{1, 1, 1, 0}
	pass = true
	for i := range r2 {
		if r2[i] != exp2[i] {
			pass = false
		}
	}
	if pass {
		fmt.Println("PASS Test 2")
	} else {
		fmt.Println("FAIL Test 2")
	}
}
